map2loop/sample_storage.py

```diff
@@ -0,0 +1,268 @@
+from abc import ABC, abstractmethod
+from .m2l_enums import Datatype, SampleType, StateType
+from .sampler import SamplerDecimator, SamplerSpacing, Sampler
+import beartype
+from typing import TYPE_CHECKING
+
+if TYPE_CHECKING:
+    from .project import Project
+
+
+class AccessStorage(ABC):
+    def __init__(self):
+        self.storage_label = "AccessStorageAbstractClass"
+
+    def type(self):
+        return self.storage_label
+
+    @beartype.beartype
+    @abstractmethod
+    def store(self, sampletype: SampleType, data):
+        pass
+
+    @beartype.beartype
+    @abstractmethod
+    def check_state(self, sampletype: SampleType):
+        pass
+
+    @abstractmethod
+    def load(self, sampletype: SampleType):
+        pass
+
+    @beartype.beartype
+    @abstractmethod
+    def process(self, sampletype: SampleType):
+        pass
+
+    @beartype.beartype
+    @abstractmethod
+    def reprocess(self, sampletype: SampleType):
+        pass
+
+    @beartype.beartype
+    @abstractmethod
+    def __call__(self, sampletype: SampleType):
+        pass
+
+
+class SampleSupervisor(AccessStorage):
+    """
+    The SampleSupervisor class is responsible for managing the samples and samplers in the project.
+    It extends the AccessStorage abstract base class.
+
+    Attributes:
+        storage_label (str): The label of the storage.
+        samples (list): A list of samples.
+        samplers (list): A list of samplers.
+        sampler_dirtyflags (list): A list of flags indicating if the sampler has changed.
+        dirtyflags (list): A list of flags indicating the state of the data, sample or sampler.
+        project (Project): The project associated with the SampleSupervisor.
+        map_data (MapData): The map data associated with the project.
+    """
+
+    def __init__(self, project: "Project"):
+        """
+        The constructor for the SampleSupervisor class.
+
+        Args:
+            project (Project): The Project class associated with the SampleSupervisor.
+        """
+
+        self.storage_label = "SampleSupervisor"
+        self.samples = [None] * len(SampleType)
+        self.samplers = [None] * len(SampleType)
+        self.sampler_dirtyflags = [True] * len(SampleType)
+        self.dirtyflags = [True] * len(StateType)
+        self.set_default_samplers()
+        self.project = project
+        self.map_data = project.map_data
+
+    def type(self):
+        return self.storage_label
+
+    def set_default_samplers(self):
```

Comment:
Shouldn't default be chosen with appropriate parameterisation given the map/dataset?